File: ai_service/symptom_phrases_loader.py
"""Phrase-level symptom index from symptom_phrases.csv and WV expansions."""

from __future__ import annotations

import csv
import re
from functools import lru_cache
from pathlib import Path
from typing import Any

_NLP = Path(__file__).resolve().parent.parent / "data" / "nlp"
_SOURCES = (
    _NLP / "symptom_phrases.csv",
    _NLP / "hiligaynon_wv_expansion.csv",
    _NLP / "hiligaynon_reproductive_expansion.csv",
    _NLP / "hiligaynon_combinatorial_phrases.csv",
    _NLP / "hiligaynon_conditions_combinatorial.csv",
    _NLP / "step6_triage_exemplars.csv",
    _NLP / "symptom_phrases_seed.csv",
)


def normalize(text: str) -> str:
    text = (text or "").lower().strip()
    text = re.sub(r"[^a-z0-9\s\-]", " ", text)
    return re.sub(r"\s+", " ", text).strip()


@lru_cache(maxsize=1)
def phrase_index() -> dict[str, dict[str, Any]]:
    index: dict[str, dict[str, Any]] = {}
    for path in _SOURCES:
        if not path.is_file():
            continue
        with path.open(encoding="utf-8", newline="") as f:
            reader = csv.DictReader(f)
            for raw in reader:
                hil = (raw.get("hiligaynon_term") or "").strip()
                eng = (raw.get("english_term") or raw.get("english_translation") or "").strip()
                if not hil or not eng:
                    continue
                key = normalize(hil)
                if key in index:
                    continue
                cat = (raw.get("medical_category") or "general").strip()
                sev = (raw.get("severity") or "Low").strip()
                tri = (raw.get("triage_level") or "routine").strip()
                index[key] = {
                    "hiligaynon_term": hil,
                    "english_term": eng,
                    "medical_category": cat,
                    "severity": sev,
                    "triage_level": tri,
                    "body_part": _infer_body_part(hil, eng),
                    "symptom": _infer_symptom(eng),
                    "condition": eng if cat in {"infection", "injury", "trauma", "gynecologic_symptom"} else "",
                    "source": path.name,
                }
    return index


@lru_cache(maxsize=1)
def phrases_by_length() -> list[str]:
    return sorted(phrase_index().keys(), key=len, reverse=True)


def _infer_body_part(hil: str, eng: str) -> str:
    mapping = {
        "itlog": "testicle", "itlug": "testicle", "bilat": "vagina", "bilad": "vagina",
        "ari": "penis", "bayag": "scrotum", "kipay": "vulva", "singit": "groin",
        "mata": "eye", "kamot": "hand", "tiil": "foot", "ulo": "head",
    }
    low = hil.lower()
    for hil_part, eng_part in mapping.items():
        if re.search(rf"\b{re.escape(hil_part)}\b", low):
            return eng_part
    eng_low = eng.lower()
    for part in ("vagina", "penis", "testicle", "scrotum", "vulva", "groin", "eye", "hand", "foot"):
        if part in eng_low:
            return part
    return ""


def _infer_symptom(eng: str) -> str:
    low = eng.lower()
    for token in ("infection", "bleeding", "swelling", "pain", "itching", "lump", "wound", "redness"):
        if token in low:
            return token
    return "symptom"


def lookup_phrase(text: str) -> dict[str, Any] | None:
    return phrase_index().get(normalize(text))


def scan_phrases(text: str) -> list[dict[str, Any]]:
    """Longest non-overlapping phrase matches (static CSV + combinatorial engine)."""
    if not text:
        return []
    working = normalize(text)
    matches = _scan_static_phrases(working)
    try:
        from phrase_combinatorial_engine import match_phrases

        for entry in match_phrases(text):
            phrase = entry.get("matched_phrase") or ""
            if phrase and not any(m.get("matched_phrase") == phrase for m in matches):
                matches.append(entry)
    except ImportError:
        pass
    matches.sort(key=lambda m: m["span"][0])
    return matches


def _phrase_candidates_from_text(working: str, max_phrase_words: int = 8) -> list[str]:
    """Only check phrase n-grams present in text against the index (not all 28k phrases)."""
    index = phrase_index()
    words = working.split()
    if not words:
        return []
    found: set[str] = set()
    word_count = len(words)
    for i in range(word_count):
        for j in range(i + 1, min(word_count, i + max_phrase_words) + 1):
            phrase = " ".join(words[i:j])
            if phrase in index:
                found.add(phrase)
    return sorted(found, key=len, reverse=True)


def _scan_static_phrases(working: str) -> list[dict[str, Any]]:
    occupied = [False] * max(len(working), 1)
    index = phrase_index()
    matches: list[dict[str, Any]] = []

    for phrase in _phrase_candidates_from_text(working):
        pattern = re.compile(r"(?<!\w)" + re.escape(phrase) + r"(?!\w)")
        for match in pattern.finditer(working):
            start, end = match.start(), match.end()
            if any(occupied[start:end]):
                continue
            entry = index.get(phrase)
            if not entry:
                continue
            for i in range(start, end):
                occupied[i] = True
            matches.append({**entry, "matched_phrase": phrase, "span": [start, end]})
    matches.sort(key=lambda m: m["span"][0])
    return matches


def clear_cache() -> None:
    phrase_index.cache_clear()
    phrases_by_length.cache_clear()

File: app/core/SymptomPhrasesLoader.php
<?php
/**
 * Symptom phrase index for phrase-first NLP matching.
 */

final class SymptomPhrasesLoader
{
    /** @var array<string, array<string, mixed>>|null */
    private static ?array $index = null;

    /** @var list<string>|null */
    private static ?array $byLength = null;

    /** @var list<string> */
    private static array $sources = [
        '/data/nlp/symptom_phrases.csv',
        '/data/nlp/hiligaynon_wv_expansion.csv',
        '/data/nlp/hiligaynon_reproductive_expansion.csv',
        '/data/nlp/hiligaynon_combinatorial_phrases.csv',
        '/data/nlp/hiligaynon_conditions_combinatorial.csv',
        '/data/nlp/hiligaynon_conditions.csv',
        '/data/nlp/hiligaynon_symptoms.csv',
        '/data/nlp/step6_triage_exemplars.csv',
        '/data/nlp/symptom_phrases_seed.csv',
    ];

    /** @return array<string, array<string, mixed>> */
    public static function phraseIndex(): array
    {
        if (self::$index !== null) {
            return self::$index;
        }

        self::$index = [];
        foreach (self::$sources as $rel) {
            $path = BASE_PATH . $rel;
            if (!is_readable($path)) {
                continue;
            }
            $handle = fopen($path, 'r');
            if ($handle === false) {
                continue;
            }
            $header = fgetcsv($handle);
            while (($row = fgetcsv($handle)) !== false) {
                $data = array_combine(
                    array_map(static fn ($h) => strtolower(trim((string) $h)), $header ?: []),
                    array_map(static fn ($v) => trim((string) $v), $row)
                ) ?: [];
                $hil = (string) ($data['hiligaynon_term'] ?? '');
                $eng = (string) ($data['english_term'] ?? $data['english_translation'] ?? '');
                if ($hil === '' || $eng === '') {
                    continue;
                }
                $key = self::normalize($hil);
                if (isset(self::$index[$key])) {
                    continue;
                }
                self::$index[$key] = [
                    'hiligaynon_term'  => $hil,
                    'english_term'     => $eng,
                    'medical_category' => (string) ($data['medical_category'] ?? 'general'),
                    'severity'         => (string) ($data['severity'] ?? 'Low'),
                    'triage_level'     => (string) ($data['triage_level'] ?? 'routine'),
                    'body_part'        => self::inferBodyPart($hil, $eng),
                    'symptom'          => self::inferSymptom($eng),
                    'source'           => basename($path),
                ];
            }
            fclose($handle);
        }

        return self::$index;
    }

    /** @return list<string> */
    public static function phrasesByLength(): array
    {
        if (self::$byLength !== null) {
            return self::$byLength;
        }
        $keys = array_keys(self::phraseIndex());
        usort($keys, static fn (string $a, string $b): int => strlen($b) <=> strlen($a));
        self::$byLength = $keys;

        return self::$byLength;
    }

    /** @return array<string, mixed>|null */
    public static function lookupPhrase(string $text): ?array
    {
        return self::phraseIndex()[self::normalize($text)] ?? null;
    }

    /** @return list<array<string, mixed>> */
    public static function scanPhrases(string $text): array
    {
        $working = self::normalize($text);
        if ($working === '') {
            return [];
        }

        $matches = self::scanStaticPhrases($working);
        $comb = PhraseCombinatorialEngine::matchPhrases($text);
        foreach ($comb as $entry) {
            $phrase = (string) ($entry['matched_phrase'] ?? '');
            if ($phrase === '') {
                continue;
            }
            $dup = false;
            foreach ($matches as $existing) {
                if (($existing['matched_phrase'] ?? '') === $phrase) {
                    $dup = true;
                    break;
                }
            }
            if (!$dup) {
                $matches[] = $entry;
            }
        }
        usort($matches, static fn (array $a, array $b): int => ($a['span'][0] ?? 0) <=> ($b['span'][0] ?? 0));

        return $matches;
    }

    /** @return list<array<string, mixed>> */
    private static function scanStaticPhrases(string $working): array
    {
        $occupied = array_fill(0, strlen($working), false);
        $matches = [];
        foreach (self::phrasesByLength() as $phrase) {
            if (!preg_match_all('/(?<!\w)' . preg_quote($phrase, '/') . '(?!\w)/iu', $working, $m, PREG_OFFSET_CAPTURE)) {
                continue;
            }
            $entry = self::phraseIndex()[$phrase] ?? null;
            if ($entry === null) {
                continue;
            }
            foreach ($m[0] as [$_, $start]) {
                $end = $start + strlen($phrase);
                $overlap = false;
                for ($i = $start; $i < $end; $i++) {
                    if ($occupied[$i] ?? false) {
                        $overlap = true;
                        break;
                    }
                }
                if ($overlap) {
                    continue;
                }
                for ($i = $start; $i < $end; $i++) {
                    $occupied[$i] = true;
                }
                $matches[] = array_merge($entry, ['matched_phrase' => $phrase, 'span' => [$start, $end]]);
            }
        }
        usort($matches, static fn (array $a, array $b): int => ($a['span'][0] ?? 0) <=> ($b['span'][0] ?? 0));

        return $matches;
    }

    private static function normalize(string $text): string
    {
        $text = strtolower(trim($text));
        $text = preg_replace('/[^a-z0-9\s\-]/', ' ', $text) ?? $text;

        return trim(preg_replace('/\s+/', ' ', $text) ?? $text);
    }

    private static function inferBodyPart(string $hil, string $eng): string
    {
        $map = ['itlog' => 'testicle', 'bilat' => 'vagina', 'ari' => 'penis', 'bayag' => 'scrotum'];
        $low = strtolower($hil);
        foreach ($map as $k => $v) {
            if (str_contains($low, $k)) {
                return $v;
            }
        }
        $engLow = strtolower($eng);
        foreach (['vagina', 'penis', 'testicle', 'scrotum', 'vulva', 'groin'] as $part) {
            if (str_contains($engLow, $part)) {
                return $part;
            }
        }

        return '';
    }

    private static function inferSymptom(string $eng): string
    {
        $low = strtolower($eng);
        foreach (['infection', 'bleeding', 'swelling', 'pain', 'itching', 'lump', 'wound'] as $t) {
            if (str_contains($low, $t)) {
                return $t;
            }
        }

        return 'symptom';
    }
}
